src/lsp.ts

```diff
@@ -0,0 +1,180 @@
+import * as Admzip from "adm-zip";
+import * as which from "which";
+import * as https from "https";
+import * as crypto from "crypto";
+import * as fs from "fs";
+import * as os from "os";
+import * as path from "path";
+import * as vscode from "vscode";
+import { ExtensionContext, WorkspaceConfiguration, workspace } from "vscode";
+import { Executable, LanguageClient, LanguageClientOptions, ServerOptions, TransportKind } from "vscode-languageclient/node";
+
+export abstract class LanguageServerClient {
+  config: WorkspaceConfiguration;
+  ls: LanguageClient | null = null;
+  executableName: string;
+  context: ExtensionContext;
+  private clientOptions: LanguageClientOptions = {
+    documentSelector: ["meson", { "scheme": "file", language: "meson" }]
+  };
+  abstract repoURL: string;
+  abstract get debugExe(): Executable
+  abstract get runExe(): Executable
+
+  constructor(executableName: string, context: ExtensionContext) {
+    this.executableName = executableName;
+    this.context = context;
+    this.config = workspace.getConfiguration("mesonbuild");
+    const serverModule = this.languageServerPath;
+
+    if (serverModule == null)
+      return;
+    this.startLanguageServer();
+  }
+
+  abstract supportsSystem(os: string, arch: string): boolean
+
+  abstract get downloadInfo(): [url: string, hash: string] | null;
+
+  restart(): void {
+    this.dispose();
+
+    if (this.languageServerPath == null)
+      return;
+    this.startLanguageServer();
+  }
+
+  private startLanguageServer() {
+    const serverOptions: ServerOptions = {
+      run: this.runExe,
+      debug: this.debugExe,
+      transport: TransportKind.stdio
+    };
+
+    this.ls = new LanguageClient("Swift-MesonLSP", "Meson Language Server", serverOptions, this.clientOptions, true);
```

Comment:
What is the second argument to this constructor?

Reply:
It's this string here:

![image](https://user-images.githubusercontent.com/63594136/233178083-b8de0051-2aad-40bb-b051-dec48efb2008.png)

Reply:
Oh. Should the whole Meson extension use the output channel? To me that would make sense. What are your thoughts?

Reply:
But maybe that is not possible? If not, then this is fine :).

Reply:
There is e.g. extra output from my language server that would e.g. help to diagnose issues and some other stuff

![image](https://user-images.githubusercontent.com/63594136/233179506-ff9e6878-e127-4805-9f45-06df53631d1e.png)


![image](https://user-images.githubusercontent.com/63594136/233179555-9632a6fa-d327-4e56-bb19-11fcf96bed6b.png)

Reply:
Is there any way to just add a prefix to log messages coming from your LS?

Reply:
I don't know, as I'm not that well-versed with the VSCode API

Reply:
Just leave as is, and we can explore this more later since it really isn't that important.